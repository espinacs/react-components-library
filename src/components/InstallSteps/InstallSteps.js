import React, { Fragment } from 'react';
import { PropTypes } from 'prop-types';
import { Steps, Button, Card, message } from 'antd';
import styles from './styles.module.css';

const { Step } = Steps;

class InstallSteps extends React.Component {
  state = {
    current: this.props.initialStep,
  };

  prev = () => {
    this.setState(({ current }) => ({ current: current - 1 }));
  };
  set = (num) => {
    this.setState(() => ({ current: num }));
  };
  next = () => {
    this.setState(({ current }) => ({ current: current + 1 }));
  };

  render() {
    const { current } = this.state;
    const { steps, onFinish } = this.props;
    return (steps && steps.length === 0) ? null : (
      <Fragment>
        <Card>
          <Steps progressDot current={current} className={styles.stepsMain}>
            {steps.map((item, idx) =>
              <Step key={item.title} title={item.title} onClick={() => this.set(idx)} />)}
          </Steps>
        </Card>

        <Card className={styles.stepsContent}>
          <div data-testid="step-content">
            <div className={styles.stepContentLeft} >
             <img src={steps[current].imgSrc} />
            </div>
            <div className={styles.stepContentRight} >
              {steps[current].content}
            </div>
          </div>

          <div className={styles.stepsAction}>
            { current > 0 &&
              <Button
                data-testid="button-prev"
                onClick={this.prev}
              >
                Previous
              </Button>
            }
            { current < steps.length - 1 &&
              <Button
                type="primary"
                data-testid="button-next"
                onClick={this.next}
              >
                Next
              </Button>
            }
            { current === steps.length - 1 &&
              <Button
                type="primary"
                data-testid="button-done"
                onClick={onFinish}
              >
                Done
              </Button>
            }
          </div>
        </Card>
      </Fragment>
    );
  }
}

InstallSteps.propTypes = {
  steps: PropTypes.arrayOf(PropTypes.shape({
    title: PropTypes.string,
    content: PropTypes.string,
    imgSrc: PropTypes.string,
  })),
  initialStep: PropTypes.number,
  onFinish: PropTypes.func,
};

InstallSteps.defaultProps = {
  steps: [],
  initialStep: 0,
  onFinish: () => message.success('Processing complete!'),
};

export default InstallSteps;
