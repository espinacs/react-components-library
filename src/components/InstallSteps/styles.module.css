.installation-stepper {
  margin: 20px;
  height: 200px;
  text-align: center;
  font-size: 40px;
  line-height: 200px;
  background: #eee;
  color: #666;
}

.steps-main {
  cursor: pointer;
}

.steps-content {
  margin-top: 10px;
}
.step-content-right {
  float: right;
  width: 70%;
}
.step-content-left {
  float: left;
  width: 30%;
}


.steps-action {
  text-align: center;
}
