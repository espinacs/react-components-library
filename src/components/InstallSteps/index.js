import InstallSteps from './InstallSteps';

export default InstallSteps;
