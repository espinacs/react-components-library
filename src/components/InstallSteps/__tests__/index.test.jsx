import React from 'react';
import { renderIntoDocument, render, cleanup } from 'react-testing-library';
import InstallSteps from '../index';


const testData = [
  {
    title: 'Extension Installed',
    content: 'A new prompt will show that the installation process is complete.',
    imgSrc: 'http://10.102.138.11:3001/assets/images/chrome/tracker-add.png',
  },
  {
    title: 'Add Extension',
    content: 'When you click the install button above, A browser prompt will appear when you click the install button',
    imgSrc: 'http://10.102.138.11:3001/assets/images/chrome/tracker-installed.png',
  },
];

afterEach(cleanup);

test('mounts empty, renders null', () => {
  const { container } = render(<InstallSteps />);
  expect(container.firstChild).toMatchSnapshot();
});

test('renders with valid props and data is displayed', () => {
  const {
    getByTestId, queryByText, getByText,
  } = renderIntoDocument(<InstallSteps steps={testData} />);
  // check content is rendered
  expect(getByTestId('step-content').textContent).toEqual(testData[0].content);
  // check previous button is not available
  expect(queryByText('Previous')).toBeNull();
  // check no done button is available
  expect(queryByText('Done')).toBeNull();
  // check next button is available
  expect(getByText('Next').textContent).toEqual('Next');
});

test('renders with valid props and moves forward', () => {
  const {
    getByTestId, queryByText, getByText,
  } = renderIntoDocument(<InstallSteps steps={testData} />);
  try {
    getByTestId('button-next').click();
  } catch (e) {
    throw new Error(e);
  }
  // check proper content is rendered
  expect(getByTestId('step-content').textContent).toEqual(testData[1].content);
  // check all buttons are available except for next
  expect(getByText('Previous').textContent).toEqual('Previous');
  expect(queryByText('Next')).toBeNull();
  expect(getByText('Done').textContent).toEqual('Done');
});

test('renders with valid props and moves backward', () => {
  const {
    getByTestId, queryByText, getByText,
  } = renderIntoDocument(<InstallSteps initialStep={1} steps={testData} />);
  getByTestId('button-prev').click();

  // check we are back to square 0
  expect(getByTestId('step-content').textContent).toEqual(testData[0].content);
  expect(queryByText('Previous')).toBeNull();
  expect(queryByText('Done')).toBeNull();
  expect(getByText('Next').textContent).toEqual('Next');
});

test('renders with valid props and gets to the end', () => {
  const {
    getByTestId, queryByText, getByText,
  } = renderIntoDocument(<InstallSteps initialStep={1} steps={testData} />);
  getByTestId('button-done').click();

  // check we are at final step
  expect(getByTestId('step-content').textContent).toEqual(testData[1].content);
  expect(getByText('Previous').textContent).toEqual('Previous');
  expect(queryByText('Next')).toBeNull();
  expect(getByText('Done').textContent).toEqual('Done');
});

test('renders with valid props and calls onFinish when gets to the end', () => {
  const handleFinish = jest.fn()
  const { getByTestId } = renderIntoDocument(<InstallSteps
    steps={testData}
    initialStep={1}
    onFinish={handleFinish}
  />);
  getByTestId('button-done').click();
  // check function onFinish has been called
  expect(handleFinish).toHaveBeenCalledTimes(1);
});


test('renders with valid props and uses top navigation', () => {
  const {
    getByTestId, queryByText, getByText,
  } = renderIntoDocument(<InstallSteps initialStep={1} steps={testData} />);

  getByText(testData[0].title).click();
  // check we are back to square 0
  expect(getByTestId('step-content').textContent).toEqual(testData[0].content);
  expect(getByText('Next').textContent).toEqual('Next');
  expect(queryByText('Previous')).toBeNull();
  expect(queryByText('Done')).toBeNull();
});

