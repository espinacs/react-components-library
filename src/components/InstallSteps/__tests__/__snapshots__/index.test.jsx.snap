// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`mounts empty, renders null 1`] = `null`;
